cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
   friend array<Y> const_array_cast(const array<U>&);
 
   public:
-    explicit array(std::int64_t element_count)
-        : data_(new T[element_count]), // TODO: Use custom oneDAL allocator
-          size_(element_count) {}
-
-    explicit array(const detail::shared<T>& data, std::int64_t size)
-        : data_(data),
-          size_(size) {}
+    array()
+        : data_owned_ptr_(nullptr),
+          size_(0),
+          capacity_(0) {}
+
+    explicit array(std::int64_t element_count, T element = {})
+        : data_owned_ptr_(new T[element_count], // TODO: Use custom oneDAL allocator
+                [](T* obj){ delete[] obj; }),
+          data_(data_owned_ptr_.get()),
+          size_(element_count),
+          capacity_(element_count) {
+        for (std::int64_t i = 0; i < size_; i++) {
+            (*this)[i] = element;
+        }
+    }
 
     template <typename Deleter>
     explicit array(T* data, std::int64_t size, Deleter&& deleter)
```

Comment:
It would be beneficial to have the template parameter different from `T` for constructor:
```cpp
template< class Y > 
explicit shared_ptr( Y* ptr );
```

Reply:
Probably we can create another free function for this. In my opinion, relying on implicit pointer conversion inside the constructor is a bad idea for `array` - it is designed to hold multiple elements, while `shared_ptr` design is mostly influenced by ownership onto one element. Let's consider the following code:
```cpp
class A {};
class B : public A {};

...

auto b_size = 1000;
B* b_data = generateBs(1000);
array<A> arr { b_data, b_size, empty_deleter() };
```
Here user gets undefined behavior while the code looks good. This looks like the error-prone approach, since we force user to be careful and write something like this:
```cpp
A** b_data = { new B(), ..., new B() };
array<A**> arr { b_data, 1000 };
```